src/dispatcher.rs

```diff
@@ -131,26 +116,19 @@ impl Dispatcher {
         }
     }
 
-    fn register_callout(&self, token_id: u32) {
-        if self
-            .callouts
-            .borrow_mut()
-            .insert(token_id, self.active_id.get())
-            .is_some()
-        {
-            panic!("duplicate token_id")
-        }
-    }
-
     fn on_create_context(&self, context_id: u32, root_context_id: u32) {
         if root_context_id == 0 {
-            self.create_root_context(context_id)
-        } else if self.new_http_stream.get().is_some() {
-            self.create_http_context(context_id, root_context_id);
-        } else if self.new_stream.get().is_some() {
-            self.create_stream_context(context_id, root_context_id);
+            self.create_root_context(context_id);
+        } else if let Some(root_context) = self.roots.borrow().get(&root_context_id) {
+            match root_context.get_type() {
+                ContextType::HttpContext => self.create_http_context(context_id, root_context_id),
+                ContextType::StreamContext => {
+                    self.create_stream_context(context_id, root_context_id)
+                }
+                ContextType::RootContext => panic!("missing ContextType on root_context"),
+            }
```

Comment:
One more thing. I think it would be useful if we could retain backwards compatibility, and merge this without breaking all existing plugins, even if there aren't too many in the wild right now.

One way to do this is to fallback to the standalone constructors in case `root_context.get_type()` returns `ContextType::RootContext` or `None`, and perhaps revert changes to `examples/http_auth_random.rs` to make sure that the old code path still works.

Reply:
+1 to remain backwards-compatible, but I think we should then deprecate the old way to create contexts and remove it in 0.2.0

Reply:
that would also allow us to remove `NoopRoot` in 0.2.0

Reply:
The "old way" is fine for plugins that don't have configuration (e.g. where everything is already in the code), and the "new way" can double the code size with unnecessary boilerplate.

But it's something to consider for sure.